rpc/flipt/request.go

```diff
@@ -49,6 +49,12 @@ type Request struct {
 	Status    Status   `json:"status"`
 }
 
+func WithNoNamespace() func(*Request) {
+	return func(r *Request) {
+		r.Namespace = ""
+	}
```

Comment:
im confused about what the difference is between using this and what you did previously by setting `WithNamespace("")` in the `ListNamespaceRequest`

Reply:
@markphelps  It would be nice if you double check it with me. Burned once already... 

`WithNamespace` doesn't allow to set empty string for some reason
https://github.com/flipt-io/flipt/blob/2ca2273a0f1e3181befb2f2fedfe548fdc519866/rpc/flipt/request.go#L52-L58

As `WithNamespace` is a public func, I can't remove condition without breaking changes for others. So I've created a new option `WithNoNamespace`.

Reply:
I think we are the only ones who use this function as its only used internally to enforce authz, so I think its fine to remove the check for empty string in the function body

Reply:
@markphelps  Could you please help me how to move forward? The integration tests has a case where the empty namespace should fallback to default namespace. Should I skip this case for authz integration tests or should I continue with `WithNoNamespace` option and keep all the integration test cases?

Reply:
i guess ideally we should continue again with the `WithNoNamespace` option to keep the integration tests. we should likely add one to make sure that they can list namespaces still if `WithNoNamespace` is used

Reply:
I am stuck. Integration tests for authz don't have calls with ListNamespaceRequest as I see. What is going on? Where is my blind spot? Why namespaced viewer could be affected by this change?

Reply:
@erka I will take a look this evening

Reply:
so the policy for the IT is coming from here: https://github.com/flipt-io/flipt/blob/main/build/testing/integration.go#L704-L839

```
default allow = false

allow if {
    input.authentication.metadata["is_bootstrap"] == "true"
}

allow if {
	some rule in has_rules

	permit_string(rule.resource, input.request.resource)
	permit_slice(rule.actions, input.request.action)
	permit_string(rule.namespace, input.request.namespace)
}

allow if {
	some rule in has_rules

	permit_string(rule.resource, input.request.resource)
	permit_slice(rule.actions, input.request.action)
	not rule.namespace
}

has_rules contains rules if {
	some role in data.roles
	role.name == input.authentication.metadata["io.flipt.auth.role"]
	rules := role.rules[_]
}

has_rules contains rules if {
	some role in data.roles
	role.name == input.authentication.metadata["io.flipt.auth.k8s.serviceaccount.name"]
	rules := role.rules[_]
}
```

and the failing test is here: https://github.com/flipt-io/flipt/blob/main/build/testing/integration/authn/auth.go#L145

so there is some reason that the policy is matching, but that doesn't make sense because we are not passing in a role in the request,  so `has_rules` should be returning the empty set, but it must be returning some rules which makes me think the `""` role is being matched somewhere in the policy

I think our rego policy for the authz test is not compatible with the change of allowing no namespace in the request

Reply:
A wild guess...  if tests are using rest api and `getNamespace` with empty namespace, the url probably will be `/api/v1/namespaces/` and grpc-gateway could remove the last flash and interpret that a call to list namespaces.